lib/spack/spack/package.py

```diff
@@ -983,28 +977,42 @@ def module(self):
                           fromlist=[self.__class__.__name__])
 
 
-    def setup_dependent_environment(self, module, spec, dependent_spec):
-        """Called before the install() method of dependents.
+    def environment_modifications(self, dependent_spec):
```

Comment:
I think this should just take an EnvironmentModifications() as a parameter, and not require the package author to know how to create one.  e.g.:

```python
def setup_environment(self, env):
    env.prepend_path(...)
    env.append_path(...)
```

Also, I think the dependent_spec aspect is kind of confusing.  It may be that we need two functions:
```python
def setup_environment(self, env):
    """hook in a package to set up *its own* environment for anything that wants to use it."""
    # most normal module stuff goes here

def setup_dependent_environment(self, env, dependent_spec):
    """hook in a package that allows it to set up the environment specifically for *dependent* packages."""
    # stuff that python might do, like add a site-packages directory to the `PYTHONPATH` on behalf
    # of an extension, should go here.
```

I think this separates the things that need to go into a normal module from the things that need to go into the *dependent's* module.  When you generate the module file for a dependent, you would probably want to include env modifications from its *dependencies'* `setup_dependent_environment` functions and from *its* `setup_environment` function.

Does that seem reasonable?